src/plugins/mpris/extension.cpp

```diff
@@ -0,0 +1,203 @@
+    qDebug("[%s] Extension initialized", name_);
+}
+
+
+
+/** ***************************************************************************/
+MPRIS::Extension::~Extension() {
+    qDebug("[%s] Finalize extension", name_);
+    // Do sth.
+    qDebug("[%s] Extension finalized", name_);
+}
+
+
+
+/** ***************************************************************************/
+QWidget *MPRIS::Extension::widget(QWidget *parent) {
+    if (widget_.isNull()) {
+        widget_ = new ConfigWidget(parent);
+    }
+    return widget_;
+}
+
+
+
+/** ***************************************************************************/
+void MPRIS::Extension::setupSession() {
+    // Querying the DBus to list all available services
+    QDBusMessage findPlayerMsg = QDBusMessage::createMethodCall("org.freedesktop.DBus", "/", "org.freedesktop.DBus", "ListNames");
+    QDBusMessage response = QDBusConnection::sessionBus().call(findPlayerMsg);
+
+    // Do some error checking
+    if (response.type() == QDBusMessage::ReplyMessage) {
+        QList<QVariant> args = response.arguments();
+        if (args.length() == 1) {
+            QVariant arg = args.at(0);
+            if (!arg.isNull() && arg.isValid()) {
+                QStringList names = arg.toStringList();
+                if (!names.isEmpty()) {
+                    // No errors
+
+                    // Filter all mpris capable
+                    names = names.filter(QRegExp("org\\.mpris\\.MediaPlayer2\\.(.*)"));
+                    mediaPlayers.clear();
+                    for (QString busid : names) {
+                        // And add their player object to the list
+                        mediaPlayers.append(Player(busid));
+                    }
+
+
+                } else {
+                    qCritical("[%s] DBus error: Argument is either not type of QStringList or is empty!", name_);
+                }
+            } else {
+                qCritical("[%s] DBus error: Reply argument not valid or null!", name_);
+            }
+        } else {
+            qCritical("[%s] DBus error: Expected 1 argument for DBus reply. Got %d", name_, args.length());
+        }
+    } else {
+        qCritical("[%s] DBus error: %s", name_, response.errorMessage().toStdString().c_str());
+    }
+}
+
+
+
+/** ***************************************************************************/
+void MPRIS::Extension::teardownSession() {
+}
+
+
+
+/** ***************************************************************************/
+bool MPRIS::Extension::pairsort(QPair<int, QString> left, QPair<int, QString> right) {
+    return left.first < right.first;
+}
+
+
+
+/** ***************************************************************************/
+void MPRIS::Extension::handleQuery(shared_ptr<Query> query) {
+    // Do not proceed if there are no players running. Why would you even?
+    if (mediaPlayers.isEmpty())
+        return;
+
+    // Filter all applicable commands
+    QString q = query->searchTerm();
+    QRegExp rx(q);
```

Comment:
Note that the default behaviour for the search is a "prefix search". Substring searches are messy, since every letter will appear somewhere in an item display text, especially vowals.  if all extensions would use it, a single letter would ouput virtually all items. Further a user would not enter "se" if he wanted to pau***se*** the media player. always keep usablility in mind.  last but not least regexes are pretty expensive.  just use Qstring::startswith.